src/bank.rs

```diff
@@ -484,12 +484,24 @@ impl Bank {
             }).collect()
     }
 
+    // Return true if the given account is being used by its program or if owned by pubkey that signed this transaction.
+    // Note: does not authorize if owned by a different program, even if signed.
+    fn is_debit_authorized(
+        account_index: usize,
+        instruction_program_id: &Pubkey,
+        account_program_id: &Pubkey,
+    ) -> bool {
+        instruction_program_id == account_program_id
+            || SystemProgram::check_id(&account_program_id) && account_index == 0
```

Comment:
Why this case?   Just because I signed a transaction from a system program account doesn't mean the program may exfiltrate those tokens.  I would use SystemProgram::Move if I want to send the program some of my tokens

Reply:
It's wrong to view a signed program as potentially malicious.  The program is required to spend tokens exactly as described by the program. The program and its operational semantics must be immutable. This is the "code is law" philosophy. All validators ensure those tokens are spend it exactly the way described by the program. It's only in the case of a sender that doesn't understand the program, that the sender *does* trust SystemProgram::Move, and that the sender *doesn't* need the program to update its userdata (i.e. erc20 transfer), should the user consider using SystemProgram::Move to minimize the risk of accidental loss of tokens.

Reply:
I don't think it's wrong at all to view a program as potentially malicious.  That happens all the time in other app ecosystems.  Why would ours be different.

I don't understand what you are trying to solve by relaxing this restriction.   If you want to remove `SystemProgram::Assign`, let's do it -- I actually can't think of much use for it right now beyond a pinch of sugar over `SystemProgram::Create` + `SystemProgram::Move`.

Reply:
The problem I'm trying to solve is to make it so a program can once again do useful things with a single instruction, let's say it's `Budget::new()`.

Reply:
If you want to view a program as potentially malicious, that's fine, move your tokens to a separate account. But if you *do* understand the program, then you should be able to go ahead and grant it access to the account tokens and userdata, and be confident it'll be modified in exactly the way the transaction authorizes. No need to make client code jump through hoops.